pkg/serviceaccount/tokens_controller.go

```diff
@@ -240,8 +244,16 @@ func (e *TokensController) secretDeleted(obj interface{}) {
 		return
 	}
 
-	if _, err := e.removeSecretReferenceIfNeeded(serviceAccount, secret.Name); err != nil {
-		glog.Error(err)
+	for i := 1; i <= NumServiceAccountRemoveReferenceRetries; i++ {
```

Comment:
The list can't be determined by inspecting secrets. The service account's list of secrets includes not only the token secrets, but any other secret you want pods referencing the service account to be able to mount.